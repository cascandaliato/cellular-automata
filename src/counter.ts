class Counter<E extends number | string> {
  private map = new Map<E, number>();

  constructor(iterable: Iterable<E> = []) {
    for (const element of iterable) this.add(element);
  }

  add(element: E): Counter<E> {
    this.map.set(element, (this.map.get(element) || 0) + 1);
    return this;
  }

  remove(element: E): Counter<E> {
    this.map.set(element, (this.map.get(element) || 0) - 1);
    return this;
  }

  get(element: E): number {
    if (!this.map.has(element)) return 0;
    return this.map.get(element) as number;
  }
}

export { Counter };
