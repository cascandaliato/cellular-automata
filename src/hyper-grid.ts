import { TupleOf } from './tuple-of';

const SEP = ';';

class HyperGrid<N extends number, V>
  implements Iterable<[TupleOf<N, number>, V]> {
  private map = new Map<string, V>();

  constructor(private defaultValue: V) {}

  get(coords: TupleOf<N, number>): V {
    return this.map.get(coords.join(SEP)) ?? this.defaultValue;
  }

  set(coords: TupleOf<N, number>, value: V) {
    if (value === this.defaultValue) {
      this.map.delete(coords.join(SEP));
    } else {
      this.map.set(coords.join(SEP), value);
    }
  }

  delete(coords: TupleOf<N, number>) {
    this.map.delete(coords.join(SEP));
  }

  *[Symbol.iterator]() {
    for (const [key, value] of this.map) {
      yield [key.split(SEP).map(Number), value] as [TupleOf<N, number>, V];
    }
  }

  get size(): number {
    return this.map.size;
  }
}

export { HyperGrid };
