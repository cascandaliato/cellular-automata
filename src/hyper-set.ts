import { TupleOf } from './tuple-of';

const SEP = ';';

class HyperSet<N extends number> implements Iterable<TupleOf<N, number>> {
  private set = new Set<string>();

  has(coords: TupleOf<N, number>): boolean {
    return this.set.has(coords.join(SEP));
  }

  add(coords: TupleOf<N, number>): HyperSet<N> {
    this.set.add(coords.join(SEP));
    return this;
  }

  remove(coords: TupleOf<N, number>): boolean {
    return this.delete(coords);
  }

  delete(coords: TupleOf<N, number>): boolean {
    return this.set.delete(coords.join(SEP));
  }

  toggle(coords: TupleOf<N, number>): boolean {
    if (this.has(coords)) {
      this.delete(coords);
      return false;
    } else {
      this.add(coords);
      return true;
    }
  }

  *[Symbol.iterator]() {
    for (const coords of this.set) {
      yield coords.split(SEP).map(Number) as TupleOf<N, number>;
    }
  }

  toArray() {
    return [...this.set].map(coords =>
      coords.split(SEP).map(Number)
    ) as TupleOf<N, number>[];
  }

  get size(): number {
    return this.set.size;
  }
}

export { HyperSet };
