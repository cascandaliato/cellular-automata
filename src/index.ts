export { Automaton } from './automaton';
export * as Limits from './limits';
export * as Neighbors from './neighbors';
export * from './presets';
