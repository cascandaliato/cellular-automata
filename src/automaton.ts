import { Counter } from './counter';
import { HyperGrid } from './hyper-grid';
import { HyperSet } from './hyper-set';
import { TupleOf } from './tuple-of';

type Coords<N extends number> = TupleOf<N, number>;

type Entry<N extends number, STATE extends string> = {
  coords: Coords<N>;
  state: STATE;
};

type Entries<N extends number, STATE extends string> = Entry<N, STATE>[];

type AutomatonConfig<N extends number, STATE extends string> = {
  cell: {
    defaultState: STATE;
    states: Record<
      STATE,
      (coords: Coords<N>, neighbors: Entries<N, STATE>) => STATE
    >;
  };
  initialState: Entries<N, STATE>;
  getNeighbors(coords: Coords<N>): Coords<N>[];
  isWithinLimits(coords: Coords<N>): boolean;
};

class ConfigurableAutomaton<
  DIMENSIONS extends number,
  STATES extends string[]
> {
  private grid: HyperGrid<DIMENSIONS, STATES[number]>;
  private generation: number = 0;
  private history: Entries<DIMENSIONS, STATES[number]>[] = [];

  constructor(private config: AutomatonConfig<DIMENSIONS, STATES[number]>) {
    this.grid = new HyperGrid<DIMENSIONS, STATES[number]>(
      config.cell.defaultState
    );
    for (const cell of config.initialState)
      this.grid.set(cell.coords, cell.state);
  }

  evolve(n: number = 1) {
    const getNeighborsWithinLimits = (coords: Coords<DIMENSIONS>) =>
      this.config.getNeighbors(coords).filter(this.config.isWithinLimits);
    for (let gen = 1; gen <= n; gen++) {
      let expanded = new HyperSet<DIMENSIONS>();
      [...this.grid].forEach(([coords]) => {
        expanded.add(coords);
        getNeighborsWithinLimits(coords).forEach(nCoords =>
          expanded.add(nCoords)
        );
      });
      const newGrid = new HyperGrid<DIMENSIONS, STATES[number]>(
        this.config.cell.defaultState
      );
      [...expanded].forEach(coords => {
        const neighborsLocal: Entries<
          DIMENSIONS,
          STATES[number]
        > = getNeighborsWithinLimits(coords).map(nCoords => ({
          coords: nCoords,
          state: this.grid.get(nCoords),
        }));
        const newState = this.config.cell.states[this.grid.get(coords)](
          coords,
          neighborsLocal
        );
        newGrid.set(coords, newState);
      });
      this.generation += 1;
      this.history.push(this.toEntries());
      this.grid = newGrid;
    }

    return this;
  }

  toEntries(): Entries<DIMENSIONS, STATES[number]> {
    return [...this.grid].map(([coords, state]) => ({
      coords,
      state,
    }));
  }

  countByState(): Counter<STATES[number]> {
    return new Counter<STATES[number]>(
      this.toEntries().map(({ state }) => state)
    );
  }

  print(): void {
    const entries = this.toEntries();
    if (entries.length === 0) return;

    const dims = entries[0].coords.length;

    const minMax = this.toEntries()
      .map(({ coords }) => coords)
      .reduce(
        (acc, coords) => {
          for (let d = 0; d < coords.length; d++) {
            acc[d][0] = Math.min(acc[d][0], coords[d]);
            acc[d][1] = Math.max(acc[d][1], coords[d]);
          }
          return acc;
        },
        Array.from({ length: dims }, () => [+Infinity, -Infinity])
      );

    if (dims === 1) {
      let line = '';
      for (let i = minMax[0][0]; i <= minMax[0][1]; i++) {
        line += this.grid.get([i] as TupleOf<DIMENSIONS, number>);
      }
      console.log(line);
      console.log();
    } else if (dims === 2) {
      console.log(`gen=${this.generation}`);
      for (let r = minMax[0][0]; r <= minMax[0][1]; r++) {
        let row = '';
        for (let c = minMax[1][0]; c <= minMax[1][1]; c++) {
          row += this.grid.get([r, c] as TupleOf<DIMENSIONS, number>);
        }
        console.log(row);
      }
      console.log();
    }
  }
}

const Automaton = <DIMENSIONS extends number, STATES extends string[]>(
  config: AutomatonConfig<DIMENSIONS, STATES[number]>
) => new ConfigurableAutomaton<DIMENSIONS, STATES>(config);

export { Automaton, AutomatonConfig, ConfigurableAutomaton, Coords, Entries };
