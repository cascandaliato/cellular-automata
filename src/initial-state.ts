import { Entries } from './automaton';
import { HyperRectangle } from './utils';

const fromMatrix = <STATE extends string>(
  matrix: HyperRectangle<STATE>
): Entries<number, STATE> => {
  const entries: Entries<number, STATE> = [];
  matrix.forEach((value: STATE | HyperRectangle<STATE>, idx: number) => {
    if (Array.isArray(value)) {
      for (const { coords, state } of fromMatrix(value)) {
        entries.push({ coords: [idx, ...coords], state });
      }
    } else {
      entries.push({ coords: [idx], state: value });
    }
  });
  return entries;
};

const fromHyperRectangle = fromMatrix;

export { fromHyperRectangle, fromMatrix };
