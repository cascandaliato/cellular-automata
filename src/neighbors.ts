import { range } from 'lodash';
import { Coords } from './automaton';
import { TupleOf } from './tuple-of';
import { product } from './utils';

const getNeighborsFromDirections = <DIMENSIONS extends number>(
  directions: TupleOf<DIMENSIONS, number>[]
) => (coords: Coords<DIMENSIONS>): Coords<DIMENSIONS>[] =>
  directions.map(d => coords.map((n, idx) => n + d[idx]) as Coords<DIMENSIONS>);

const mooreDirections = (n: number): TupleOf<typeof n, number>[] =>
  product([0, 1, -1], n).filter(p => p.some(v => v !== 0));

const moore = (dimensions: number) =>
  getNeighborsFromDirections<typeof dimensions>(mooreDirections(dimensions));

const vonNeumannDirections = (
  dimensions: number,
  distance: number,
  residue: number = distance
): TupleOf<typeof dimensions, number>[] => {
  if (dimensions === 0) return residue === 0 ? [[]] : [];

  return range(residue !== 0 ? -residue : 0, residue + 1).flatMap(r =>
    vonNeumannDirections(
      dimensions - 1,
      distance,
      residue - Math.abs(r)
    ).map(d => [...d, r])
  );
};

// https://en.wikipedia.org/wiki/Von_Neumann_neighborhood
const vonNeumann = (dimensions: number, distance: number = 1) =>
  getNeighborsFromDirections<typeof dimensions>(
    range(1, distance + 1).flatMap(d => vonNeumannDirections(dimensions, d))
  );

export {
  moore,
  moore as surrounding,
  vonNeumann,
  vonNeumann as byManhattanDistance,
};
