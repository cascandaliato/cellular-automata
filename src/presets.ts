import { merge } from 'lodash';
import { Automaton, AutomatonConfig, Coords, Entries } from './automaton';
import * as InitialState from './initial-state';
import * as Limits from './limits';
import * as Neighbors from './neighbors';
import { TupleOf } from './tuple-of';
import { countEntriesByState, DeepPartial, HyperRectangle } from './utils';

const Conway = (
  dimensions: number,
  dead: string,
  alive: string,
  initialState: Entries<typeof dimensions, typeof dead | typeof alive> = [],
  overrides: DeepPartial<
    AutomatonConfig<typeof dimensions, typeof dead | typeof alive>
  > = {}
) => {
  type DIMENSIONS = typeof dimensions;
  type DEAD = typeof dead;
  type ALIVE = typeof alive;
  type STATE = DEAD | ALIVE;

  return Automaton<DIMENSIONS, DEAD | ALIVE>(
    merge(
      {
        cell: {
          defaultState: dead,
          states: {
            [alive]: (
              _: TupleOf<DIMENSIONS, number>,
              neighbors: {
                coords: TupleOf<DIMENSIONS, number>;
                state: STATE;
              }[]
            ) =>
              [2, 3].includes(countEntriesByState(neighbors, alive))
                ? alive
                : dead,
            [dead]: (
              _: TupleOf<DIMENSIONS, number>,
              neighbors: {
                coords: TupleOf<DIMENSIONS, number>;
                state: STATE;
              }[]
            ) => (countEntriesByState(neighbors, alive) === 3 ? alive : dead),
          } as Record<
            STATE,
            (
              coords: Coords<DIMENSIONS>,
              neighbors: Entries<DIMENSIONS, STATE>
            ) => STATE
          >, // https://github.com/microsoft/TypeScript/issues/13948
        },
        initialState,
        getNeighbors: Neighbors.moore(dimensions),
        isWithinLimits: () => true,
      },
      overrides
    )
  );
};

Conway.fromMatrix = (
  dimensions: number,
  dead: string,
  alive: string,
  matrix: HyperRectangle<typeof dead | typeof alive> = []
) =>
  Conway(
    dimensions,
    dead,
    alive,
    (InitialState.fromMatrix<typeof dead | typeof alive>(
      matrix
    ) as unknown) as Entries<typeof dimensions, typeof dead | typeof alive>
  );

const ConwayInfinite = Conway;

const Conway2D = (
  dead: string,
  alive: string,
  initialState: Entries<2, typeof dead | typeof alive> = [],
  overrides: DeepPartial<AutomatonConfig<2, typeof dead | typeof alive>> = {}
) => Conway(2, dead, alive, initialState as Entries<number, string>, overrides);

Conway2D.fromMatrix = (
  dead: string,
  alive: string,
  matrix: HyperRectangle<typeof dead | typeof alive> = []
) => Conway.fromMatrix(2, dead, alive, matrix);

const ConwayFinite = (
  limits: number[] | [number, number][],
  dead: string,
  alive: string,
  initialState: Entries<typeof limits.length, typeof dead | typeof alive> = [],
  overrides: DeepPartial<
    AutomatonConfig<typeof limits.length, typeof dead | typeof alive>
  > = {}
) =>
  Conway(limits.length, dead, alive, initialState, {
    ...overrides,
    isWithinLimits: Limits.fromArray(limits),
  });

ConwayFinite.fromMatrix = (
  limits: number[] | [number, number][],
  dead: string,
  alive: string,
  matrix: HyperRectangle<typeof dead | typeof alive> = []
) =>
  ConwayFinite(
    limits,
    dead,
    alive,
    (InitialState.fromMatrix<typeof dead | typeof alive>(
      matrix
    ) as unknown) as Entries<typeof limits.length, typeof dead | typeof alive>
  );

export { Conway, Conway2D, ConwayInfinite, ConwayFinite };
