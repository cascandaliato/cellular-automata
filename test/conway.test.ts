import { vonNeumann } from '../src/neighbors';
import { Conway } from '../src/presets';

describe("Conway's Game of Life", () => {
  test('After one iteration', () => {
    const gol = Conway.fromMatrix(2, '.', '#', [
      ['.', '#', '.'],
      ['.', '.', '#'],
      ['#', '#', '#'],
    ]);

    gol.evolve(1);
    expect(gol.toEntries()).toMatchSnapshot();
  });
});

describe('Neighbors', () => {
  test("VonNeumann's neighborhood (1d, distance 1)", () => {
    expect(vonNeumann(1, 1)([1])).toMatchSnapshot();
  });

  test("VonNeumann's neighborhood (1d, distance 2)", () => {
    expect(vonNeumann(1, 2)([1])).toMatchSnapshot();
  });

  test("VonNeumann's neighborhood (2d, distance 1)", () => {
    expect(vonNeumann(2, 2)([1, 1])).toMatchSnapshot();
  });
  test("VonNeumann's neighborhood (2d, distance 2)", () => {
    expect(vonNeumann(2, 2)([1, 1])).toMatchSnapshot();
  });

  test("VonNeumann's neighborhood (3d, distance 1)", () => {
    expect(vonNeumann(3, 1)([1, 1, 0])).toMatchSnapshot();
  });
  test("VonNeumann's neighborhood (3d, distance 2)", () => {
    expect(vonNeumann(3, 2)([1, 1, 0])).toMatchSnapshot();
  });
});
