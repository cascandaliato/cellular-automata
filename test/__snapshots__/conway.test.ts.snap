// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Conway's Game of Life After one iteration 1`] = `
Array [
  Object {
    "coords": Array [
      1,
      2,
    ],
    "state": "#",
  },
  Object {
    "coords": Array [
      1,
      0,
    ],
    "state": "#",
  },
  Object {
    "coords": Array [
      2,
      2,
    ],
    "state": "#",
  },
  Object {
    "coords": Array [
      2,
      1,
    ],
    "state": "#",
  },
  Object {
    "coords": Array [
      3,
      1,
    ],
    "state": "#",
  },
]
`;

exports[`Neighbors VonNeumann's neighborhood (1d, distance 1) 1`] = `
Array [
  Array [
    0,
  ],
  Array [
    2,
  ],
]
`;

exports[`Neighbors VonNeumann's neighborhood (1d, distance 2) 1`] = `
Array [
  Array [
    0,
  ],
  Array [
    2,
  ],
  Array [
    -1,
  ],
  Array [
    3,
  ],
]
`;

exports[`Neighbors VonNeumann's neighborhood (2d, distance 1) 1`] = `
Array [
  Array [
    1,
    0,
  ],
  Array [
    0,
    1,
  ],
  Array [
    2,
    1,
  ],
  Array [
    1,
    2,
  ],
  Array [
    1,
    -1,
  ],
  Array [
    0,
    0,
  ],
  Array [
    2,
    0,
  ],
  Array [
    -1,
    1,
  ],
  Array [
    3,
    1,
  ],
  Array [
    0,
    2,
  ],
  Array [
    2,
    2,
  ],
  Array [
    1,
    3,
  ],
]
`;

exports[`Neighbors VonNeumann's neighborhood (2d, distance 2) 1`] = `
Array [
  Array [
    1,
    0,
  ],
  Array [
    0,
    1,
  ],
  Array [
    2,
    1,
  ],
  Array [
    1,
    2,
  ],
  Array [
    1,
    -1,
  ],
  Array [
    0,
    0,
  ],
  Array [
    2,
    0,
  ],
  Array [
    -1,
    1,
  ],
  Array [
    3,
    1,
  ],
  Array [
    0,
    2,
  ],
  Array [
    2,
    2,
  ],
  Array [
    1,
    3,
  ],
]
`;

exports[`Neighbors VonNeumann's neighborhood (3d, distance 1) 1`] = `
Array [
  Array [
    1,
    1,
    -1,
  ],
  Array [
    1,
    0,
    0,
  ],
  Array [
    0,
    1,
    0,
  ],
  Array [
    2,
    1,
    0,
  ],
  Array [
    1,
    2,
    0,
  ],
  Array [
    1,
    1,
    1,
  ],
]
`;

exports[`Neighbors VonNeumann's neighborhood (3d, distance 2) 1`] = `
Array [
  Array [
    1,
    1,
    -1,
  ],
  Array [
    1,
    0,
    0,
  ],
  Array [
    0,
    1,
    0,
  ],
  Array [
    2,
    1,
    0,
  ],
  Array [
    1,
    2,
    0,
  ],
  Array [
    1,
    1,
    1,
  ],
  Array [
    1,
    1,
    -2,
  ],
  Array [
    1,
    0,
    -1,
  ],
  Array [
    0,
    1,
    -1,
  ],
  Array [
    2,
    1,
    -1,
  ],
  Array [
    1,
    2,
    -1,
  ],
  Array [
    1,
    -1,
    0,
  ],
  Array [
    0,
    0,
    0,
  ],
  Array [
    2,
    0,
    0,
  ],
  Array [
    -1,
    1,
    0,
  ],
  Array [
    3,
    1,
    0,
  ],
  Array [
    0,
    2,
    0,
  ],
  Array [
    2,
    2,
    0,
  ],
  Array [
    1,
    3,
    0,
  ],
  Array [
    1,
    0,
    1,
  ],
  Array [
    0,
    1,
    1,
  ],
  Array [
    2,
    1,
    1,
  ],
  Array [
    1,
    2,
    1,
  ],
  Array [
    1,
    1,
    2,
  ],
]
`;
